File: tests/mentions.test.ts
import { describe, expect, test, mock } from 'bun:test';
import { resolveMentions, type Mentionable } from '../src/utils/mentions';

const mentionables: Mentionable[] = [
  {
    name: 'Wayne Smith',
    firstName: 'Wayne',
    sgid: 'wayne-sgid-123',
    avatarUrl: '/6103476/users/1/avatar',
    userId: '1',
  },
  {
    name: 'Wahab Khurram',
    firstName: 'Wahab',
    sgid: 'wahab-sgid-456',
    avatarUrl: '/6103476/users/2/avatar',
    userId: '2',
  },
  {
    name: 'Faraz Ali',
    firstName: 'Faraz',
    sgid: 'faraz-sgid-789',
    avatarUrl: '/6103476/users/3/avatar',
    userId: '3',
  },
];

describe('resolveMentions', () => {
  test('@firstName resolves to mention HTML', () => {
    const html = '<p>Hey @Wayne, check this</p>';
    const result = resolveMentions(html, mentionables);
    expect(result).toContain('sgid="wayne-sgid-123"');
    expect(result).toContain('content-type="application/vnd.actiontext.mention"');
    expect(result).toContain('title="Wayne Smith"');
    expect(result).not.toContain('@Wayne');
  });

  test('@"Full Name" resolves to mention HTML', () => {
    const html = '<p>Hey @"Wayne Smith", check this</p>';
    const result = resolveMentions(html, mentionables);
    expect(result).toContain('sgid="wayne-sgid-123"');
    expect(result).toContain('title="Wayne Smith"');
    expect(result).not.toContain('@"Wayne Smith"');
  });

  test('multiple mentions in one string', () => {
    const html = '<p>@Wayne and @Wahab please review</p>';
    const result = resolveMentions(html, mentionables);
    expect(result).toContain('sgid="wayne-sgid-123"');
    expect(result).toContain('sgid="wahab-sgid-456"');
    expect(result).not.toContain('@Wayne');
    expect(result).not.toContain('@Wahab');
  });

  test('unresolved @Unknown stays as plain text', () => {
    const originalWarn = console.warn;
    const warnings: string[] = [];
    console.warn = (msg: string) => warnings.push(msg);

    const html = '<p>Hey @Unknown, check this</p>';
    const result = resolveMentions(html, mentionables);
    expect(result).toContain('@Unknown');
    expect(result).not.toContain('action-text-attachment');
    expect(warnings.some(w => w.includes('@Unknown'))).toBe(true);

    console.warn = originalWarn;
  });

  test('email addresses are NOT treated as mentions', () => {
    const html = '<p>Email me at user@example.com</p>';
    const result = resolveMentions(html, mentionables);
    expect(result).toBe(html);
    expect(result).not.toContain('action-text-attachment');
  });

  test('case-insensitive matching', () => {
    const html = '<p>Hey @wayne, check this</p>';
    const result = resolveMentions(html, mentionables);
    expect(result).toContain('sgid="wayne-sgid-123"');
  });

  test('ambiguous first name warns and skips', () => {
    const ambiguousMentionables: Mentionable[] = [
      ...mentionables,
      {
        name: 'Wayne Jones',
        firstName: 'Wayne',
        sgid: 'wayne2-sgid-999',
        avatarUrl: '/6103476/users/99/avatar',
        userId: '99',
      },
    ];

    const originalWarn = console.warn;
    const warnings: string[] = [];
    console.warn = (msg: string) => warnings.push(msg);

    const html = '<p>Hey @Wayne, check this</p>';
    const result = resolveMentions(html, ambiguousMentionables);
    // Should NOT replace — ambiguous
    expect(result).toContain('@Wayne');
    expect(result).not.toContain('action-text-attachment');
    expect(warnings.some(w => w.includes('Ambiguous'))).toBe(true);

    console.warn = originalWarn;
  });

  test('ambiguous first name can be resolved with @"Full Name"', () => {
    const ambiguousMentionables: Mentionable[] = [
      ...mentionables,
      {
        name: 'Wayne Jones',
        firstName: 'Wayne',
        sgid: 'wayne2-sgid-999',
        avatarUrl: '/6103476/users/99/avatar',
        userId: '99',
      },
    ];

    const html = '<p>Hey @"Wayne Smith", check this</p>';
    const result = resolveMentions(html, ambiguousMentionables);
    expect(result).toContain('sgid="wayne-sgid-123"');
    expect(result).not.toContain('wayne2-sgid-999');
  });

  test('@ inside HTML tags is not treated as a mention', () => {
    const html = '<a href="mailto:user@example.com">contact</a>';
    const result = resolveMentions(html, mentionables);
    expect(result).toBe(html);
  });

  test('empty mentionables returns html unchanged', () => {
    const html = '<p>@Wayne test</p>';
    const result = resolveMentions(html, []);
    expect(result).toBe(html);
  });

  test('mention HTML includes avatar img with correct attributes', () => {
    const html = '<p>@Wayne</p>';
    const result = resolveMentions(html, mentionables);
    expect(result).toContain('src="/6103476/users/1/avatar"');
    expect(result).toContain('width="48"');
    expect(result).toContain('height="48"');
  });

  test('mention at start of paragraph', () => {
    const html = '<p>@Faraz please look at this</p>';
    const result = resolveMentions(html, mentionables);
    expect(result).toContain('sgid="faraz-sgid-789"');
    expect(result).toContain('title="Faraz Ali"');
  });
});

File: internal/commands/mentions_test.go
package commands

import (
	"fmt"
	"strings"
	"testing"

	"github.com/basecamp/fizzy-cli/internal/client"
)

const samplePromptUsersHTML = `
<lexxy-prompt-item search="Wayne Smith WS me" sgid="wayne-sgid-123">
  <template type="menu">
    <img aria-hidden="true" title="Wayne Smith" src="/123/users/u1/avatar" width="48" height="48" />
    Wayne Smith
  </template>
  <template type="editor">
    <img aria-hidden="true" title="Wayne Smith" src="/123/users/u1/avatar" width="48" height="48" />
    Wayne
  </template>
</lexxy-prompt-item>
<lexxy-prompt-item search="Bushra Gul BG" sgid="bushra-sgid-456">
  <template type="menu">
    <img aria-hidden="true" title="Bushra Gul" src="/123/users/u2/avatar" width="48" height="48" />
    Bushra Gul
  </template>
  <template type="editor">
    <img aria-hidden="true" title="Bushra Gul" src="/123/users/u2/avatar" width="48" height="48" />
    Bushra
  </template>
</lexxy-prompt-item>
<lexxy-prompt-item search="Kennedy K" sgid="kennedy-sgid-789">
  <template type="menu">
    <img aria-hidden="true" title="Kennedy" src="/123/users/u3/avatar" width="48" height="48" />
    Kennedy
  </template>
  <template type="editor">
    <img aria-hidden="true" title="Kennedy" src="/123/users/u3/avatar" width="48" height="48" />
    Kennedy
  </template>
</lexxy-prompt-item>
`

func newMentionMockClient() *MockClient {
	m := NewMockClient()
	m.GetHTMLResponse = &client.APIResponse{
		StatusCode: 200,
		Body:       []byte(samplePromptUsersHTML),
	}
	return m
}

func TestParseMentionUsers(t *testing.T) {
	users := parseMentionUsers([]byte(samplePromptUsersHTML))

	if len(users) != 3 {
		t.Fatalf("expected 3 users, got %d", len(users))
	}

	tests := []struct {
		idx       int
		firstName string
		fullName  string
		sgid      string
	}{
		{0, "Wayne", "Wayne Smith", "wayne-sgid-123"},
		{1, "Bushra", "Bushra Gul", "bushra-sgid-456"},
		{2, "Kennedy", "Kennedy", "kennedy-sgid-789"},
	}

	for _, tt := range tests {
		u := users[tt.idx]
		if u.FirstName != tt.firstName {
			t.Errorf("user[%d] FirstName = %q, want %q", tt.idx, u.FirstName, tt.firstName)
		}
		if u.FullName != tt.fullName {
			t.Errorf("user[%d] FullName = %q, want %q", tt.idx, u.FullName, tt.fullName)
		}
		if u.SGID != tt.sgid {
			t.Errorf("user[%d] SGID = %q, want %q", tt.idx, u.SGID, tt.sgid)
		}
		if u.AvatarSrc == "" {
			t.Errorf("user[%d] AvatarSrc is empty", tt.idx)
		}
	}
}

func TestParseMentionUsersEmpty(t *testing.T) {
	users := parseMentionUsers([]byte(""))
	if len(users) != 0 {
		t.Errorf("expected 0 users from empty HTML, got %d", len(users))
	}
}

func TestResolveMentions(t *testing.T) {
	tests := []struct {
		name           string
		input          string
		shouldContain  []string
		shouldNotMatch []string // substrings that should NOT appear
	}{
		{
			name:           "no @ passthrough",
			input:          "Hello world",
			shouldContain:  []string{"Hello world"},
			shouldNotMatch: []string{"action-text-attachment"},
		},
		{
			name:          "single mention",
			input:         "Hey @Wayne check this",
			shouldContain: []string{`sgid="wayne-sgid-123"`, `content-type="application/vnd.actiontext.mention"`, `title="Wayne Smith"`, ">Wayne<"},
		},
		{
			name:          "case insensitive",
			input:         "Hey @wayne check this",
			shouldContain: []string{`sgid="wayne-sgid-123"`},
		},
		{
			name:          "multiple mentions",
			input:         "@Wayne and @Bushra please review",
			shouldContain: []string{`sgid="wayne-sgid-123"`, `sgid="bushra-sgid-456"`},
		},
		{
			name:           "email not treated as mention",
			input:          "Contact user@example.com",
			shouldContain:  []string{"user@example.com"},
			shouldNotMatch: []string{"action-text-attachment"},
		},
		{
			name:           "unresolved mention stays as text",
			input:          "Hey @Unknown person",
			shouldContain:  []string{"@Unknown"},
			shouldNotMatch: []string{"action-text-attachment"},
		},
		{
			name:          "mention at start of text",
			input:         "@Kennedy can you look?",
			shouldContain: []string{`sgid="kennedy-sgid-789"`},
		},
		{
			name:          "mention after newline",
			input:         "First line\n@Wayne second line",
			shouldContain: []string{`sgid="wayne-sgid-123"`},
		},
		{
			name:          "single name user",
			input:         "Hey @Kennedy",
			shouldContain: []string{`sgid="kennedy-sgid-789"`, `title="Kennedy"`},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			resetMentionCache()
			mock := newMentionMockClient()
			result := resolveMentions(tt.input, mock)

			for _, s := range tt.shouldContain {
				if !strings.Contains(result, s) {
					t.Errorf("result should contain %q\ngot: %s", s, result)
				}
			}
			for _, s := range tt.shouldNotMatch {
				if strings.Contains(result, s) {
					t.Errorf("result should NOT contain %q\ngot: %s", s, result)
				}
			}
		})
	}
}

func TestResolveMentionsAPIError(t *testing.T) {
	resetMentionCache()
	mock := NewMockClient()
	mock.GetHTMLError = fmt.Errorf("server error")

	// Should return text unchanged on error
	input := "Hey @Wayne"
	result := resolveMentions(input, mock)
	if result != input {
		t.Errorf("expected unchanged text on error, got: %s", result)
	}
}

func TestResolveMentionsCaching(t *testing.T) {
	resetMentionCache()
	mock := newMentionMockClient()

	// First call fetches
	resolveMentions("@Wayne", mock)
	if len(mock.GetHTMLCalls) != 1 {
		t.Errorf("expected 1 GetHTML call, got %d", len(mock.GetHTMLCalls))
	}

	// Second call uses cache
	resolveMentions("@Bushra", mock)
	if len(mock.GetHTMLCalls) != 1 {
		t.Errorf("expected still 1 GetHTML call (cached), got %d", len(mock.GetHTMLCalls))
	}
}
